source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+}
+
+uint64_t GzipFilterConfig::windowBitsUint(Protobuf::uint32 window_bits) {
+  return (window_bits > 0 ? window_bits : DefaultWindowBits) | GzipHeaderValue;
+}
+
+GzipFilter::GzipFilter(const GzipFilterConfigSharedPtr& config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  skip_compression_ = !isAcceptEncodingAllowed(headers);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding() && isEtagAllowed(headers)) {
+
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    for (const auto token :
+         StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", true)) {
+      const auto values = StringUtil::splitToken(token, ";");
+      const auto value = StringUtil::trim(values.at(0));
+      const auto q_value =
+          values.size() > 1 ? StringUtil::trim(values.at(1)) : absl::string_view{nullptr, 0};
```

Comment:
Per above comment, if more than one ; is expected, isn't it possible that the q-value may be in values[2]?

Reply:
Only one "q" per Accept-Encoding header value. 

https://tools.ietf.org/html/rfc7231#page-41

Reply:
Hmmm...in that reference I see
```
     Accept-Encoding: gzip;q=1.0, identity; q=0.5, *;q=0
```

And anyway the q= might not be in value[1] but in some other value.

Reply:
Sorry J, but I can't see how `q=` could not be in the index 1 if it exists and assuming that I check for empty values of course. 

Given ``` Accept-Encoding: gzip;q=1.0, identity; q=0.5, *;q=0```

We first split by `,` which gives us the following:
token[0] = `gzip;q=1.0`
token[1] = `identity; q=0.5`
token[2] = `*;q=0`

```
 for (const auto token : 
        StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", true)) { ... }
```

Then we loop over each token and split by `;` which give us the following:
values[0] = `gzip`, values[1] = `q=1.0`
values[1] = `identity`, values[1] = `q=0.5`
values[2] = `*`, values[1] = `q=0`

```
const auto values = StringUtil::splitToken(token, ";");
```

If the initial token conforms with the RFC and a value split by `,` is accompanied of a Qvalue, which means, can be split by `;`, `q={}` will always be in the index 1.

Reply:
Hey no worries, it doesn't matter. I've already changed it to use crop instead :) 
```
const auto value = StringUtil::trim(StringUtil::cropRight(token, ";", false));      
const auto q_value = StringUtil::trim(StringUtil::cropLeft(token, ";", false));
```